storage/src/store/mod.rs

```diff
@@ -0,0 +1,990 @@
+        let Some(old_loc) = self.get_loc(&key).await? else {
+            // Key does not exist, so this is a no-op.
+            return Ok(());
+        };
+
+        let Some(mut cursor) = self.snapshot.get_mut(&key) else {
+            return Ok(());
+        };
+
+        // Delete the old location from the snapshot.
+        while let Some(loc) = cursor.next() {
+            if *loc == old_loc {
+                cursor.delete();
+                break;
+            }
+        }
+
+        drop(cursor);
+
+        self.apply_op(Operation::Delete(key)).await.map(|_| ())
+    }
+
+    /// Commits all uncommitted operations to the store, making them persistent and recoverable.
+    pub async fn commit(&mut self) -> Result<(), Error> {
+        self.raise_inactivity_floor(self.uncommitted_ops + 1)
+            .await?;
+        self.uncommitted_ops = 0;
+
+        self.sync().await?;
+        self.prune_inactive().await
+    }
+
+    /// Closes the store. Any uncommitted operations will be lost if they have not been committed
+    /// via [`Store::commit`].
+    pub async fn close(self) -> Result<(), Error> {
+        if self.uncommitted_ops > 0 {
+            warn!(
+                op_count = self.uncommitted_ops,
+                "closing store with uncommitted operations"
+            );
+        }
+
+        try_join!(self.log.close(), self.locations.close())?;
+        Ok(())
+    }
+
+    /// Destroys the store permanently, removing all persistent data associated with it.
+    ///
+    /// # Warning
+    ///
+    /// This operation is irreversible. Do not call this method unless you are sure
+    /// you want to delete all data associated with this store permanently!
+    pub async fn destroy(self) -> Result<(), Error> {
+        try_join!(self.log.destroy(), self.locations.destroy())?;
+        Ok(())
+    }
+
+    /// Returns the number of operations that have been applied to the store.
+    pub fn op_count(&self) -> u64 {
+        self.op_count
+    }
+
+    /// Syncs the active section of the log to persistent storage.
+    ///
+    /// This method ensures that all buffered data is written to disk, but unlike [`Store::commit`],
+    /// does not create a commit point.
+    ///
+    /// Use this method when you want to ensure data durability without creating a formal
+    /// transaction boundary.
+    async fn sync(&mut self) -> Result<(), Error> {
+        let current_section = self.op_count / self.log_items_per_section;
+        try_join!(self.log.sync(current_section), self.locations.sync())?;
+        Ok(())
+    }
+
+    /// Builds the database's snapshot from the log of operations. Any operations that sit above
+    /// the latest commit operation are removed.
+    ///
+    /// Returns the number of operations that were applied to the store, the oldest retained
+    /// location, and the inactivity floor location.
+    async fn build_snapshot_from_log(mut self) -> Result<Self, Error> {
```

Comment:
This method of building the snapshot is very memory inefficient at the moment; need to update this to trim committed operations as we stream the log's contents.

Reply:
(As you originally called out, I definitely think there is a world where many of the adb's use store behind the scenes but we can make those migrations in subsequent PRs)